lib/rjgit.rb

```diff
@@ -341,21 +334,116 @@ def commit(message, author, parents = nil, ref = nil, force = false)
         ru.setRefLogIdent(author.person_ident)
         ru.setRefLogMessage("commit: #{message}", false)
         res = ru.update.to_string
-        
-        # @treebuilder.object_inserter.release
+
         @current_tree = new_tree
         log = @treebuilder.log
         @treebuilder.init_log
         sha =  ObjectId.to_string(new_head)
         return res, log, sha
       end
-      
+
       def self.successful?(result)
         ["NEW", "FAST_FORWARD", "FORCED"].include?(result)
       end
-      
+
+    end
+
+    class ApplyPatchToIndex < RJGit::Plumbing::Index
+
+      import org.eclipse.jgit.patch.Patch
+      import org.eclipse.jgit.diff.DiffEntry
+
+      ADD    = org.eclipse.jgit.diff.DiffEntry::ChangeType::ADD
+      COPY   = org.eclipse.jgit.diff.DiffEntry::ChangeType::COPY
+      MODIFY = org.eclipse.jgit.diff.DiffEntry::ChangeType::MODIFY
+      DELETE = org.eclipse.jgit.diff.DiffEntry::ChangeType::DELETE
+      RENAME = org.eclipse.jgit.diff.DiffEntry::ChangeType::RENAME
+
+      # Take the result of RJGit::Porcelain.diff with options[:patch] = true and return a patch String
+      def self.diffs_to_patch(diffs)
+        result = ""
+        diffs.each {|diff| result << diff[:patch]}
+        result
+      end
+
+      def initialize(repository, patch, ref = Constants::HEAD)
+        super(repository)
+        @ref   = ref
+        @patch = Patch.new
+        @patch.parse(ByteArrayInputStream.new(patch.to_java_bytes))
+        raise_patch_apply_error unless @patch.getErrors.to_a.empty?
+        @current_tree = Commit.find_head(@jrepo, ref).tree
+      end
+
+      def commit(message, author, parents = nil, force = false)
+        super(message, author, parents, @ref, force)
+      end
+
+      def build_map
+        raise_patch_apply_error if @patch.getFiles.to_a.empty?
```

Comment:
Probably more efficient to call `List#isEmpty()` instead:
```ruby
raise_patch_apply_error unless @patch.getFiles.isEmpty()
```

Reply:
Do you mean `List.isEmpty()` (class as opposed to instance method)? So `List.isEmpty(@patch.getFiles)`?